File: zig/armstrong-numbers/armstrong_numbers.zig
const std = @import("std");

pub fn isArmstrongNumber(n: u128) bool {
    const digit_count = countDigits(n);
    var sum: u128 = 0;
    var temp: u128 = n;

    while (temp > 0) {
        const digit: u128 = temp % 10;
        sum += std.math.pow(u128, digit, digit_count);
        temp /= 10;
    }

    return sum == n;
}

fn countDigits(n: u128) u8 {
    var count: u8 = 0;
    var temp: u128 = n;
    while (temp != 0) {
        count += 1;
        temp /= 10;
    }
    return count;
}

File: lua/armstrong-numbers/armstrong-numbers.lua
local ArmstrongNumbers = {}

function ArmstrongNumbers.is_armstrong_number(number)
  local digits = {}
  local sum = 0
  local n = number
  while n > 0 do
    table.insert(digits, n % 10)
    n = math.floor(n / 10)
  end
  for _, digit in ipairs(digits) do
    sum = sum + digit ^ #digits
  end
  return sum == number
end

return ArmstrongNumbers
